### PR TITLE
Accelerate tree traversal in auto collision problems

### LINKED ISSUE #409: Improve traversal algorithm when interested only in pairs as a result
Just recording some thoughts.

The halo finding algorithm using union-find that we have right now has the callback that proceeds only if `i < j` so as to process any edge only once. The thing to notice here is this may as well be replaced by `permuted(i) < permuted(j)`, where `permuted(i)` is the application of Morton permutation. 

The implication of this is that we could skip traversal of the whole subtrees. If we use Apetrei ordering, for example, then we know that all nodes in the left subtree have index less or equal to that the index of the internal node. Thus, if the index of the internal node is lower than that of the query (assuming that queries are permuted using the same permutation as leaf nodes), we don't need to traverse that subtree. In a similar fashion, it should also be possible for Karras ordering.

Potential use cases for pairs:
- 2-points correlation algorithm
- Half-neighbor lists in MD
- Inline force computation in MD

It may also be worthwhile to examine papers on collision problems in computer graphics, as they share some of the same features.

### PR DESCRIPTION
**Pull request by dalg24**
Fix #409

### CODE REVIEW

File: src/details/ArborX_Callbacks.hpp [hunk 1]
```diff
@@ -27,6 +27,13 @@ enum class CallbackTreeTraversalControl
   normal_continuation
 };
 
+namespace Experimental
+{
+class TreeTraversalQuick
```
Comment (dalg24) [inline src/details/ArborX_Callbacks.hpp:32]: This is a placeholder until we come up with a better name.

File: src/details/ArborX_DetailsCrsGraphWrapperImpl.hpp [hunk 2]
```diff
@@ -59,6 +59,9 @@ struct InsertGenerator
   CountView _counts;
   PermutedOffset _permuted_offset;
 
+  using TreeTraversalControl =
+      detected_t<CallbackTreeTraversalControlArchetypeAlias, Callback>;
+
```
Comment (aprokop) [inline src/details/ArborX_DetailsCrsGraphWrapperImpl.hpp:64]: OK, seeing this makes me question the semantics of this kind of traversal when storing the results. What does it mean for a user to ask for a quick/pair traversal? Specifically,
- Is it deterministic?
  If it's not, the result of the second rerun may differ from the first.
- Is it balanced?
  If it's not, should a user have no expectation at all?
In general, this implicitly ties the result to the implementation details of such a traversal. If you change the details to, for example, `query_index >= node_index`, this would completely change the result a user gets. Is this something we want?

File: src/details/ArborX_DetailsTreeTraversal.hpp [hunk 3]
```diff
@@ -97,41 +100,58 @@ struct TreeTraversal<BVH, Predicates, Callback, SpatialPredicateTag>
 
     Node const *stack[64];
     Node const **stack_ptr = stack;
-    *stack_ptr++ = nullptr;
-    Node const *node = _bvh.getRoot();
+    *stack_ptr++ = nullptr;            // push sentinel value
```
Comment (aprokop) [inline src/details/ArborX_DetailsTreeTraversal.hpp:103]: Would prefer to use something than `sentinel` here, as we already use it in a different context for ropes. Something like `termination value` should work.

File: src/details/ArborX_DetailsTreeTraversal.hpp [hunk 4]
```diff
@@ -97,41 +100,58 @@ struct TreeTraversal<BVH, Predicates, Callback, SpatialPredicateTag>
...
+    Node const *node = _bvh.getRoot(); // start with root
     do
     {
-      Node const *child_left = _bvh.getNodePtr(node->left_child);
-      Node const *child_right = _bvh.getNodePtr(node->right_child);
+      int left_child_index = node->left_child;
+      int right_child_index = node->right_child;
+
+      Node const *left_child_node = _bvh.getNodePtr(left_child_index);
+      Node const *right_child_node = _bvh.getNodePtr(right_child_index);
```
Comment (aprokop) [inline src/details/ArborX_DetailsTreeTraversal.hpp:111]: Both left and right children's indices are inside the internal node. So, there's no extra memory load associated with accessing them. However, loading the children node's here seems inefficient if you do decide to prune. Right now, the main limitation of skipping this load is because we have `isLeaf()` method part of the node and not part of the hierarchy. Otherwise, it should be possible. I had to similarly do that when trying to split internal and leaf nodes.

In short, I'm not suggesting that you change it (unless you see an easy way), just ask to add a comment.

File: src/details/ArborX_DetailsTreeTraversal.hpp [hunk 5]
```diff
@@ -97,41 +100,58 @@ struct TreeTraversal<BVH, Predicates, Callback, SpatialPredicateTag>
...
+
+      bool left_child_is_leaf = left_child_node->isLeaf();
+      bool right_child_is_leaf = right_child_node->isLeaf();
+
+      bool prune_left =
+          (left_child_is_leaf && pruneLeaf(queryIndex, left_child_index)) ||
+          (!left_child_is_leaf &&
+           pruneLeftSubtree(queryIndex, left_child_index));
```
Comment (aprokop) [inline src/details/ArborX_DetailsTreeTraversal.hpp:119]: Yeah, I don't like it, but this is currently the only way. Ideally, the hierarchy would know whether something is a leaf or not based on its index, and you would simply need a single `prune` thing.

File: src/details/ArborX_DetailsTreeTraversal.hpp [hunk 6]
```diff
@@ -97,41 +100,58 @@ struct TreeTraversal<BVH, Predicates, Callback, SpatialPredicateTag>
...
+      bool prune_right =
+          (right_child_is_leaf && pruneLeaf(queryIndex, right_child_index));
 
-      bool overlap_left = predicate(_bvh.getBoundingVolume(child_left));
-      bool overlap_right = predicate(_bvh.getBoundingVolume(child_right));
+      bool overlap_left =
+          !prune_left && predicate(_bvh.getBoundingVolume(left_child_node));
+      bool overlap_right =
+          !prune_right && predicate(_bvh.getBoundingVolume(right_child_node));
```
Comment (aprokop) [inline src/details/ArborX_DetailsTreeTraversal.hpp:126]: I think this reduces readability and debuggability, and would not include prune inside overlap. Instead, I would do 
```c++
bool traverse_left = false;
if (!prune_left)
{
   bool overlap_left = predicate(_bvh.getBoundingVolume(child_left));
   ...
   traverse_left = (overlap_left && !child_left->isLeaf());
}
```

File: src/details/ArborX_DetailsTreeTraversal.hpp (hunk 5 above)
Comment (aprokop) [reply src/details/ArborX_DetailsTreeTraversal.hpp:119]: Another comment I have is that there's no particular reason to call the function `pruneLeftSubtree`, just call it `pruneSubtree`. The reason is that the function itself is only dependent on the indices of query and node, not on whether that node is a left or right subtree.

File: src/details/ArborX_DetailsTreeTraversal.hpp [hunk 7]
```diff
@@ -154,11 +174,22 @@ struct TreeTraversal<BVH, Predicates, Callback, SpatialPredicateTag>
...
         {
-          next = node->left_child;
+          int const left_child_index = node->left_child;
+          if (!pruneLeftSubtree(queryIndex, left_child_index))
+          {
+            next = left_child_index;
+          }
+          else
+          {
+            Node const *left_child_node = _bvh.getNodePtr(left_child_index);
```
Comment (aprokop) [inline src/details/ArborX_DetailsTreeTraversal.hpp:184]: This is really unfortunate, as we still pay for loading the left child. Don't know how to avoid it, though.

File: src/details/ArborX_DetailsTreeTraversal.hpp [hunk 8]
```diff
@@ -171,6 +202,41 @@ struct TreeTraversal<BVH, Predicates, Callback, SpatialPredicateTag>
...
+    return false;
+  }
+
+  template <typename T = Callback>
+  KOKKOS_FUNCTION
+      std::enable_if_t<std::is_same<T, Callback>{} && traverse_half<Callback>{},
+                       bool>
+      pruneLeftSubtree(int query_index, int left_child_index) const
+  {
+    return left_child_index <= query_index;
```
Comment (aprokop) [inline src/details/ArborX_DetailsTreeTraversal.hpp:238]: So, we need to think a bit further about what we are comparing here.

First of all, we don't know if the predicates are sorted or not. It would work with both, but the meaning is going to be different.

Second, lets say the predicates *are* sorted. Then the `queryIndex` matches the leaf node indices, so distant leaves should have distant indices, and not overlap. I think that's OK. The way the algorithm prunes then depends on the spatial curve ordering. So, then, one cannot assume that geometrically you would only get primitives from one side of a sphere, for example.

I wonder, though, what would happen if we tried to replace `pruneLeaf` with comparing stored permutation with queryIndex? One thing it would certainly achieve is that the resulting indices would be on one side of the `queryIndex`. Not sure what it means, though, but would like to try and see the results.

File: test/tstQueryTreeCallbacks.cpp [hunk 9]
```diff
@@ -228,6 +228,51 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(callback_early_exit, TreeTypeTraits,
...
+
+struct Experimental_CustomCallbackQuickTraversal
+{
+  using TreeTraversalControl = ArborX::Experimental::TreeTraversalQuick;
+  template <class Predicate, class OutputFunctor>
+  KOKKOS_FUNCTION void operator()(Predicate const &predicate, int j,
+                                  OutputFunctor const &out) const
+  {
+    int i = getData(predicate);
+    (void)i;
```
Comment (aprokop) [inline test/tstQueryTreeCallbacks.cpp:240]: What is the point of these, again?

File: test/tstQueryTreeCallbacks.cpp [hunk 10]
```diff
@@ -228,6 +228,51 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(callback_early_exit, TreeTypeTraits,
...
+      {b, b, b, b}, {0, 1, 2, 3});
+
+  Kokkos::View<int *, DeviceType> indices("indices", 0);
+  Kokkos::View<int *, DeviceType> offsets("offsets", 0);
+  tree.query(ExecutionSpace{}, predicates,
+             Experimental_CustomCallbackQuickTraversal{}, indices, offsets);
+  BOOST_TEST(
+      make_compressed_storage(offsets, indices) ==
+          make_reference_solution<int>({1, 2, 3, 2, 3, 3}, {0, 3, 5, 6, 6}),
+      tt::per_element());
```
Comment (aprokop) [inline test/tstQueryTreeCallbacks.cpp:274]: So this is really tied to the current "pair" traversal algorithm, isn't it? Would it not be better to instead check that it finds all "pairs", independent of how those are stores in the crs storage?

**Review by aprokop — COMMENTED** [review]
Should we allow finding ourself? Right now, the algorithm is "Find all pairs (a, b), a != b, dist(a,b) <= r`. But can we think of a situation where we would want to remove the second condition?

Also, while it is currently implemented as masking left subtrees, can we think of advantages of masking right subtrees, if that even possible?

**Comment by aprokop** [comment]
Something else to mention. For this algorithm to work, the ordering of the primitives and predicates **must** be the same. Otherwise, edges may be missed.

There are two ways we can guarantee this:
1. The easiest thing to guarantee that would be to store the permutation produced for primitives, and reapply it for predicates.
2. An alternative approach is to use a stable sort for sorting Morton indices (we do *not* do right now). The reason is that if there are duplicated Morton indices, their ordering is not guaranteed in non-stable sort and may depend on runtime.

File: test/tstQueryTreeCallbacks.cpp (hunk 10 above)
Comment (aprokop) [reply test/tstQueryTreeCallbacks.cpp:274]: That's fair, though I don't think it's complicated. You could transform a crs graph into pairs, impose comparison operator on pairs like
```c++
bool operator<(Kokos::pair<int, int>& p1, Kokkos::pair<int,int>& p2) {
  int i = p1.first;
  int j = p1.second;
  int k = p2.first;
  int l = p2.second;
  if (min(i,j) < min(k,l)) return true;
  else if (min(i,j) > min(k,l)) return false;
  else if (max(i,j) < max(k,l)) return true;
  else return false;
}
```
sort both results and the reference result, and then do comparison in linear order.

File: src/details/ArborX_DetailsTreeTraversal.hpp (hunk 6 above)
Comment (aprokop) [reply src/details/ArborX_DetailsTreeTraversal.hpp:126]: I don't understand you. For me, inclusion of prune boolean into overlap clouds the meaning of the overlap. Essentially, it is not longer clear later in the code whether the value of overlap is false because there is no physical overlap, or if is false because it is pruned. For me, overlap only concerns the physical aspect of overlapping two geometric objects, not the overlapping with a tree.

File: src/details/ArborX_DetailsTreeTraversal.hpp (hunk 5 above)
Comment (aprokop) [reply src/details/ArborX_DetailsTreeTraversal.hpp:119]: I disagree. The meaning of what we prune is derived from the context. The name of the function should be based on what it does, which is actually just a check that has nothing to do with whether it is a left or a right subtree. But I am not going to require a change.

File: src/details/ArborX_DetailsCrsGraphWrapperImpl.hpp (hunk 2 above)
Comment (aprokop) [reply src/details/ArborX_DetailsCrsGraphWrapperImpl.hpp:64]: Agreed. Conceptually, I'm totally fine with the pure callback, but these questions came to mind when thinking about crs.

File: src/details/ArborX_DetailsTreeTraversal.hpp (hunk 8 above)
Comment (aprokop) [reply src/details/ArborX_DetailsTreeTraversal.hpp:238]: After thinking a bit about it, there's not much we can do here. Essentially, our answer is going to be "Find all pairs (i,j) satisfying both predicate and morton_codes(i) <= morton_codes(j)". Changing the latter condition to "i < j" is not possible, as otherwise subtrees cannot be skipped.

This actually made me realize that our output in the crs graph is always in the order of increasing Morton indices, and not in the increasing order of indices. This could actually cause some slowdown in the force calculation even when a regular ArborX calls are used.

File: src/details/ArborX_DetailsTreeTraversal.hpp (hunk 6 above)
Comment (dalg24) [reply src/details/ArborX_DetailsTreeTraversal.hpp:126]: The point is not to evaluate the predicate when we determine we want to prune (or "skip" not sure what is the best name for it).
I fail to see how the code you suggest gives you more information than the one I wrote.

**Comment by aprokop** [comment]
I rebased the branch on the current master. With the recent changes, the patch is really small now.

**Comment by aprokop** [comment]
In my [previous](https://github.com/arborx/ArborX/pull/485#issuecomment-795874566) comment, I proposed two ways to address a potential issue of not having the same permutation for the leaf nodes and queries, leading to subtle bugs. I completely missed a third alternative, which is much cleaner.

Specifically, when using queries with a quick traversal callback, we should construct a permutation using the permutation indices stored in leaf nodes, and apply it to queries. This a) avoids calling unstable sort, and b) avoids storing the permutation as a separate entity in `BVH`. It should be of small cost, as it's `O(n)` operation through a single `parallel_for()` call, and should even be faster than computing Morton indices and sorting as we do right now.

@dalg24 I'm not 100% certain what's the best way to implement this. It seems like we need some extra function in `BVH` which will construct a permutation either through calling `sortObjects`, or by going through leaf nodes. This function will have to be specialized for the signature of the callback. Thoughts about where this function should live?

**Comment by aprokop** [comment]
It just occurred to me that it is possible to see this approach from a different angle. 

Consider assigning each leaf node a label, and then reducing them to internal nodes with a `max` operator (similar to #600, but we did a different operator there). Then, each internal node knows the max label of all of its children. If we assign a label to query, we can ask it to bypass all nodes where the max label is less than the query label. The only challenge would be to assign the node labels based on permutation of the leaf nodes.

I think that overall we should strive to implement a general data reduction algorithm on a tree with custom reducers. I think this would open many interesting algorithmic options, particularly custom traversal with skips.

**Comment by aprokop** [comment]
Closing in favor of some version of #801.
